File: ch2/lineards/array_algorithms.java
import java.util.*;

// This source code is not as complete as array_algorithms.cpp

class team implements Comparable<team> {
  private int id, solved, penalty;

  public team(int id, int solved, int penalty) {
    this.id = id;
    this.solved = solved;
    this.penalty = penalty;
  }

  public int compareTo(team o) {
    if (solved != o.solved) // can use this primary field to decide sorted order
      return o.solved - solved;   // ICPC rule: sort by number of problem solved
    else if (penalty != o.penalty)         // solved == o.solved, but we can use
                                       // secondary field to decide sorted order
      return penalty - o.penalty;       // ICPC rule: sort by descending penalty
    else                          // solved == o.solved AND penalty == o.penalty
      return id - o.id;                      // sort based on increasing team ID
  }

  public String toString() {
    return "id: " + id + ", solved: " + solved + ", penalty: " + penalty;
  }
}

class array_algorithms {
  public static void main(String[] args) {
    ArrayList<Integer> v = new ArrayList<>();

    v.add(10);
    v.add(7);
    v.add(2);
    v.add(15);
    v.add(4);

    // sort descending with vector
    Collections.sort(v);
    // if we want to modify comparison function, use the overloaded method: Collections.sort(List list, Comparator c);
    Collections.reverse(v);

    System.out.println(v);
    System.out.printf("==================\n");

    // shuffle the content again
    Collections.shuffle(v);
    System.out.println(v);
    System.out.printf("==================\n");

    // sort ascending
    Collections.sort(v);
    System.out.println(v);
    System.out.printf("==================\n");

    ArrayList<team> nus = new ArrayList<>();
    nus.add(new team(1, 1, 10));
    nus.add(new team(2, 3, 60));
    nus.add(new team(3, 1, 20));
    nus.add(new team(4, 3, 60));

    // without sorting, they will be ranked like this:
    for (int i = 0; i < 4; ++i)
      System.out.println(nus.get(i));

    Collections.sort(nus);              // sort using a comparison function
    System.out.printf("==================\n");
    // after sorting using ICPC rule, they will be ranked like this:
    for (int i = 0; i < 4; ++i)
      System.out.println(nus.get(i));
    System.out.printf("==================\n");

    int pos = Collections.binarySearch(v, 7);
    System.out.println("Trying to search for 7 in v, found at index = " + pos);

    pos = Collections.binarySearch(v, 77);
    System.out.println("Trying to search for 77 in v, found at index = " + pos); // output is -5 (explanation below)

    /*
    binarySearch will returns:
      index of the search key, if it is contained in the list;
      otherwise, (-(insertion point) - 1).
      The insertion point is defined as the point at which the key would be inserted into the list:
      the index of the first element greater than the key,
      or list.size(), if all elements in the list are less than the specified key.
      Note that this guarantees that the return value will be >= 0 if and only if the key is found. 
    */

    // sometimes these two useful simple macros are used
    System.out.printf("min(10, 7) = %d\n", Math.min(10, 7));
    System.out.printf("max(10, 7) = %d\n", Math.max(10, 7));
  }
}

File: ch2/lineards/array_algorithms.py
import bisect

# we will gradually convert the C++17 version to Python 3 soon

# typedef struct {
#   int id;
#   int solved;
#   int penalty;
# } team;

# bool icpc_cmp(team a, team b) {
#   if (a.solved != b.solved) // can use this primary field to decide sorted order
#     return a.solved > b.solved;   // ICPC rule: sort by number of problem solved
#   else if (a.penalty != b.penalty)       // a.solved == b.solved, but we can use
#                                        // secondary field to decide sorted order
#     return a.penalty < b.penalty;       // ICPC rule: sort by descending penalty
#   else                        // a.solved == b.solved AND a.penalty == b.penalty
#     return a.id < b.id;                      // sort based on increasing team ID
# }

def main():
    arr = [10, 7, 2, 15, 4]
    # sort descending with list
    arr.sort()
    arr.reverse()
    for val in arr:
      print(val, end=' ')                        # access the values
    print()
    print("==================")

  # shuffle(v.begin(), v.end(), default_random_engine()); // shuffle the content
  # for (auto &val : v)
  #   printf("%d ", val);
  # printf("\n");
  # printf("==================\n");
  # partial_sort(v.begin(), v.begin()+2, v.end()); // partial_sort demo
  # for (auto &val : v)
  #   printf("%d ", val);
  # printf("\n");
  # printf("==================\n");


  # // multi-field sorting example, suppose we have 4 ICPC teams
  # team nus[4] = { {1, 1, 10}, 
  #                 {2, 3, 60},
  #                 {3, 1, 20},
  #                 {4, 3, 60} };

  # // without sorting, they will be ranked like this:
  # for (int i = 0; i < 4; ++i)
  #   printf("id: %d, solved: %d, penalty: %d\n",
  #          nus[i].id, nus[i].solved, nus[i].penalty);

  # sort(nus, nus+4, icpc_cmp);           // sort using a comparison function
  # printf("==================\n");
  # // after sorting using ICPC rule, they will be ranked like this:
  # for (int i = 0; i < 4; ++i)
  #   printf("id: %d, solved: %d, penalty: %d\n",
  #          nus[i].id, nus[i].solved, nus[i].penalty);
  # printf("==================\n");

  # // there is a technique for multi-field sorting if the sort order is "standard"
  # // use pair (for 2 fields) or tuple (for >= 3 fields) in C++ and put the highest priority in front
  # typedef tuple<int, string, string> state;
  # state a = make_tuple(10, "steven", "grace");   // old way
  # state b = {7, "steven", "halim"};              // C++17 way
  # state c = {7, "steven", "felix"};
  # state d = {9, "a", "b"};
  # vector<state> test;
  # test.push_back(a);
  # test.push_back(b);
  # test.push_back(c);
  # test.push_back(d);
  # for (auto &[value, name1, name2] : test)
  #   printf("value: %d, name1 = %s, name2 = %s\n", value, name1.c_str(), name2.c_str());
  # printf("==================\n");
  # sort(test.begin(), test.end());                // use built-in comparator
  # // sorted ascending based on value, then based on name1,
  # // then based on name2, in that order!
  # for (auto &[value, name1, name2] : test)
  #   printf("value: %d, name1 = %s, name2 = %s\n", value, name1.c_str(), name2.c_str());
  # printf("==================\n");

    # binary search using lower bound
    arr.sort()                                    # sort ascending again
    i = bisect.bisect_left(arr, 7)                # found
    print("{}".format(i))

    j = bisect.bisect_left(arr, 77)               # not found
    if j == len(arr):
        # arr is of size 5 -> arr[0], arr[1], arr[2], arr[3], arr[4]
        # if lower_bound cannot find the required value,
        # it will return len(arr)
        # thus, testing whether j == len(arr)
        # can detect this "not found" issue
        print("77 not found")
    print("==================")

    # sometimes these two useful simple macros are used
    print("min(10, 7) = {}".format(min(10, 7)))
    print("max(10, 7) = {}".format(max(10, 7)))

main()
